File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> Q 
            = new PriorityQueue<>(Comparator.comparingInt(arr -> arr[0]));

        for (int i = 0; i < points.length; i++) {
            int[] point = points[i];
            int dist = point[0] * point[0] + point[1] * point[1];

            Q.offer(new int[]{dist, i});
         }

         int result[][] = new int[k][2];

         for (int i = 0; i < k; i++) {
            result[i] = points[Q.poll()[1]];
         } 

         return result;
    }
}

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.py
import heapq
from typing import List

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        # Min-heap storing (distance, index)
        heap = []
        for i, (x, y) in enumerate(points):
            dist = x * x + y * y
            heapq.heappush(heap, (dist, i))

        result = []
        for _ in range(k):
            _, idx = heapq.heappop(heap)
            result.append(points[idx])

        return result
